woocommerce/woocommerce
# Enable searching for products by SKU

### All Submissions:






### Changes proposed in this Pull Request:



Closes #28508

#### Why we cannot use `sku` parameter

Searching products by SKU has been requested in Woo mobile apps in the past (ref p7bje6-2GK-p2), and I'm giving it a try to implement this for 2022 Q1 HACK Week. Currenctly, we have a `sku` parameter, but it has some issues for the use case in mobile:

- Only products with full SKU match are returned. This is used for validating SKU in the product inventory settings in mobile, and we most likley want to keep this behavior
- `search` and `sku` conditions are joined by `AND`, thus only products with the query in both product name/content/excerpt AND sku are returned
- If we make separate requests to `/wc/v3/products?search=woo` and `/wc/v3/products?sku=woo`, the total results might not match the requested page size and requires some custom handling to support pagination

#### Proposed solution

This PR attempted to support SKU search by introducing a new parameter `search_includes_sku: bool`. When `search_includes_sku` parameter is `true`, `wc_product_meta_lookup` table is joined and a `WHERE` condition is added to also (`OR`) find products with partial SKU match in the lookup table.

It seems like in core (Products tab), partial SKU match is already included in each search query (`s`):

https://github.com/woocommerce/woocommerce/blob/aa8b7a4a6b7d21ccddaac360766c4daa396b9a98/plugins%2Fwoocommerce%2Fincludes%2Fdata-stores%2Fclass-wc-product-data-store-cpt.php#L1651

We could also include the SKU search in the existing `search` parameter, but I thought it's safer to not affect pre-existing search usage given the performance impact.

### How to test the changes in this Pull Request:

Prerequisites: the test site has at least one product with a non-empty SKU (e.g. `woo-tshirt`) that isn't also in the product name of the same product

1. Run `{{site_url}}/wp-json/wc/v3/products?search={{sku_substring}}` (e.g. `search=Woo`) --> the product in the prequisite should not be returned, only products with matched name or other core product attributes are returned
2. Run the request in the previous step with `search_includes_sku=true` --> the product in the prequisite should be returned, along with other products returned in the previous step
3. Feel free to try other SKU substrings and letter casing in the search query (e.g. `tshirt`, `Tshirt`, `Woo-tshirt`)

### Other information:




### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### FOR PR REVIEWER ONLY:
